searx/engines/packagist.py

```diff
@@ -0,0 +1,78 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# lint: pylint
+# pylint: disable=missing-function-docstring
+"""Packagist (IT)
+
+"""
+
+from json import loads
+from urllib.parse import urlencode
+from dateutil import parser
+
+import requests
+
+from searx import logger
+
+about = {
+     "website": 'https://packagist.org',
+     "wikidata_id": 'Q108311377',
+     "official_api_documentation": 'https://packagist.org/apidoc',
+     "use_official_api": True,
+     "require_api_key": False,
+     "results": 'JSON',
+ }
+
+categories = ['it']  # optional
+paging = True
+
+base_url = "https://packagist.org/"
+search_url = base_url + "search.json?{query}&per_page=25"
+
+def metadata(package_url):
+    result = {}
+    body = loads(requests.get(package_url + ".json").text)
+
+    package = body.get("package")
+    if package:
+        maintainers = []
+
+        for m in package.get("maintainers"):
+            maintainers.append(m.get("name"))
+
+        result["author"] = ", ".join(maintainers)
+        result["publishedDate"] = parser.parse(
+            package.get("time")
+        )
+
+    return result
+
+def request(query, params):
+    params['url'] = search_url.format(
+        query = urlencode(
+            dict(q = query, page = params["pageno"])
+        )
+    )
+    logger.debug("query_url --> %s", params['url'])
+    return params
+
+def response(resp):
+    '''post-response callback
+    resp: requests response object
+    '''
+    results = []
+    body = loads(resp.text)
+
+    for item in body.get("results"):
+        result = {}
+
+        url = item.get("url")
+
+        result["url"] = url
+        result["title"] = item.get("name")
+        result["content"] = item.get("description")
+
+        result.update(metadata(url))
```

Comment:
@dalf what do you think, from a http process model view, is it OK to send additional HTTP requests here in the response function?  If not, what might be an alternative.

Reply:
It is not a problem, and it is already the case of different engines:
* openstreetmap (call to wikidata = another HTTPS request)
* seznam: additional HTTPS request in the `request` function
* some others.

The only issue here is the number of requests: one per result.

All requests are sequential, so it can be slow and timeout.

Reply:
> All requests are sequential, so it can be slow and timeout.

From chapter ["Best practices when using the Packagist.org API"](https://packagist.org/apidoc):

    If you send requests in parallel, be a good citizen and do a maximum of 10 concurrent requests,
    or up to 20 if you only fetch static files. 

@dalf: do you have a suggestion to parallelize HTTP requests (with searx.network from e3c1286)?

Reply:
> do you have a suggestion to parallelize HTTP requests (with searx.network from e3c1286)?

Switch to async, only for the engines processing, so it is possible to send multiple requests and then wait the result concurrently.

Without async: 10 new threads, but 5 requests to this engines = 55 threads (one per requests, one per result for each requests). It is not workable.